src/olympia/users/models.py

```diff
@@ -637,23 +638,79 @@ def fetch_names():
         return any(n in name for n in blocked_list)
 
 
-@python_2_unicode_compatible
-class UserRestriction(ModelBase):
+class IPAddressUserRestriction(ModelBase):
     """Define restrictions for user-based data that can be used to either
     avoid a user to register, submit add-ons or similar.
     """
     id = PositiveAutoField(primary_key=True)
     ip_address = models.GenericIPAddressField(
         blank=True, null=True,
         help_text=_('Enter a valid IPv4 or IPv6 address, e.g 127.0.0.1'))
+
+    class Meta:
+        db_table = 'users_user_ip_restriction'
+
+    @classmethod
+    def allow_request(self, request):
+        remote_addr = request.META.get('REMOTE_ADDR')
+        return not (
+            IPAddressUserRestriction.objects
+            .filter(ip_address=remote_addr)
+            .exists())
+
+    @classmethod
+    def get_error_message(self, request):
+        return _(
+            'Multiple add-ons violating our policies have been'
+            ' submitted from your location. The IP address has been'
+            ' blocked.')
+
+
+class IPNetworkUserRestriction(ModelBase):
+    id = PositiveAutoField(primary_key=True)
     network = CIDRField(
         blank=True, null=True,
         help_text=_(
             'Enter a valid IPv6 or IPv6 CIDR network range, eg. 127.0.0.1/28'))
+
+    class Meta:
+        db_table = 'users_user_network_restriction'
+
+    @classmethod
+    def allow_request(self, request):
+        try:
+            remote_addr = ipaddress.ip_address(request.META.get('REMOTE_ADDR'))
+        except ValueError:
+            # If we don't have a valid ip address, let's deny
+            # TODO: Verify this is what we want…
+            return False
+
+        restrictions = IPNetworkUserRestriction.objects.all()
+
+        for restriction in restrictions:
+            if remote_addr in list(restriction.network.hosts()):
+                return False
+        return True
+
+    @classmethod
+    def get_error_message(self, request):
+        return IPAddressUserRestriction.get_error_message()
+
+
+class EmailUserRestriction(ModelBase):
```

Comment:
Similarly to what I'm suggesting above eventually this should hold both domain (w/ wildcard) and email restrictions.